File: examples/go/mega_flow/pipeline.go
package main

import (
	"context"
	"strings"
)

// Orchestrate — exercises Go's idiomatic flow constructs: goroutines,
// channels, defer, select, context cancellation, closures, variadic
// functions, range-over-channel, type switch.

// makeJoiner — closure factory returning a reducer that joins tokens.
func makeJoiner(sep string) func(acc, tok string) string {
	return func(acc, tok string) string {
		if acc == "" {
			return tok
		}
		return acc + sep + tok
	}
}

func collectExtras(values ...string) []string {
	return values
}

// tokenize — generator goroutine that streams tainted tokens over a channel.
func tokenize(ctx context.Context, cmd string) <-chan string {
	out := make(chan string)
	go func() {
		defer close(out)
		for _, part := range strings.Fields(cmd) {
			select {
			case <-ctx.Done():
				return
			case out <- part:
			}
		}
	}()
	return out
}

func Orchestrate(ctx context.Context, env Envelope) int {
	cmd := env.Cmd
	user := env.User
	env.Extras = collectExtras(env.Extras...)

	// range-over-channel — consume tainted tokens from the goroutine.
	var tokens []string
	for tok := range tokenize(ctx, cmd) {
		tokens = append(tokens, strings.TrimSpace(tok))
	}

	// Reduce via closure — taint rides the accumulator.
	joiner := makeJoiner(" ")
	joined := ""
	for _, t := range tokens {
		if t == "" {
			continue
		}
		joined = joiner(joined, t)
	}

	// Switch with fall-through — every arm preserves taint.
	var routed string
	var boxedKind any = env.Kind
	switch k := boxedKind.(type) {
	case Kind:
		env.Kind = k
	default:
		env.Kind = KindRun
	}
	switch env.Kind {
	case KindRun:
		routed = joined
	case KindEval:
		routed = strings.TrimSpace(joined)
	default:
		routed = joined
	}

	// defer + recover (Go's try/catch analogue) — taint survives the branch.
	var valid Envelope
	func() {
		defer func() {
			if r := recover(); r != nil {
				valid = Envelope{Kind: env.Kind, Cmd: routed, User: user, Length: len(routed), Extras: env.Extras}
			}
		}()
		if routed == "" {
			panic("empty")
		}
		valid = Envelope{Kind: env.Kind, Cmd: routed, User: user, Length: len(routed), Extras: env.Extras}
	}()

	return Persist(valid)
}

File: examples/cpp/mega_flow/pipeline.cpp
// Pipeline — exercises C++'s idiomatic flow constructs: lambdas,
// std::function, templates, range-based for, structured bindings,
// <algorithm>, try/catch, switch.
#include <algorithm>
#include <functional>
#include <numeric>
#include <sstream>
#include <stdexcept>
#include <string>
#include <tuple>
#include <vector>

#include "envelope.hpp"

int persist(Envelope env);
using TokenList = std::vector<std::string>;

// Closure factory — returns a std::function reducer that joins tokens.
static std::function<std::string(std::string, const std::string&)> make_joiner(const std::string& sep) {
    return [sep](std::string acc, const std::string& tok) -> std::string {
        if (acc.empty()) return tok;
        return acc + sep + tok;
    };
}

// Templated tokenizer — range-based iteration via stringstream.
template <typename Container>
static Container tokenize(const std::string& cmd) {
    Container out;
    std::istringstream iss(cmd);
    std::string part;
    while (iss >> part) {
        if (!part.empty()) out.push_back(part);
    }
    return out;
}

int orchestrate(Envelope env) {
    // Collect tokens via template + range-for.
    TokenList tokens = tokenize<TokenList>(env.cmd);
    for (const auto& token : tokens) {
        if (token.empty()) continue;
    }

    // std::accumulate + lambda reducer — taint rides the accumulator.
    auto joiner = make_joiner(" ");
    std::string joined = std::accumulate(
        tokens.begin(), tokens.end(), std::string{}, joiner);
    auto [routed_seed, routed_len] = std::make_tuple(joined, joined.size());
    (void)routed_len;

    // switch — every arm preserves taint.
    std::string routed;
    switch (env.kind) {
        case Kind::Run:  routed = routed_seed; break;
        case Kind::Eval: routed = routed_seed; break;
    }

    // try / catch — taint survives every branch.
    Envelope valid{env};
    try {
        if (routed.empty()) throw std::runtime_error("empty");
        valid.cmd = routed;
        valid.length = static_cast<int>(routed.size());
    } catch (const std::exception&) {
        valid.cmd = routed;
        valid.length = static_cast<int>(routed.size());
    }

    return persist(valid);
}
